Handle the exception in call-back function

iusyk opened the pull request:
There is a list of the callback functions, to be called from the kernel of Wayland
Below is the sample of callback call (see wl_proxy_create_for_global)
    wl_list_for_each(listener, &display->global_listener_list, link)
                    (*listener->handler)(display, &proxy->base, listener->data);
    
If callback function throws an exception it rather brings the crash,  because of not caught exception.
Summary: all code that potentially can throw an exception has to be wrapped in try-catch block.
The following changes are done
File Display.cpp
1.Display::sRegistryHandler:
The call 'static_cast<Display*>(data)->registryHandler(registry, id, interface, version);' wrapped
in try-catch block. In the case of the exception, the error is written in log.
2.Display::sRegistryRemove:fix that does not have a relation to the issue, but improve the 'safe' behavior
Display::sRegistryRemove: the call of wl_registry_add_listener is check against the returned value.
If result is -1 the exception is thrown.

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>

iusyk committed:
Handle the exception in call-back function

There is a list of the callback functions, to be called from
the kernel of Wayland
Below is the sample of callback call (see wl_proxy_create_for_global)
wl_list_for_each(listener, &display->global_listener_list, link)
		(*listener->handler)(display, &proxy->base, listener->data);

If callback function throws an exception it rather brings the crash,
because of not caught exception.
Summary: all code that potentially can throw an exception has to be wrapped in
try-catch block.
The following change is  done
File Display.cpp
Display::sRegistryHandler:
The call 'static_cast<Display*>(data)->registryHandler(registry, id, interface, version);' wrapped
in try-catch block. In the case of the exception, the error is written in log.

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>

arminn commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -408,8 +408,13 @@ void Display::sWaylandLog(const char* fmt, va_list arg)
+		static_cast<Display*>(data)->registryHandler(registry, id, interface, version);
+	}
+	catch(const std::exception& err){
+		LOG("Waylanf", ERROR) << err.what();
```
WaylanD

iusyk replied:
> WaylanD

fixed

lorc commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -408,8 +408,13 @@ void Display::sWaylandLog(const char* fmt, va_list arg)
+	{
+		static_cast<Display*>(data)->registryHandler(registry, id, interface, version);
+	}
+	catch(const std::exception& err){
```
Line break is missing before "{"

iusyk replied:
fixed

lorc commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -428,24 +433,22 @@ void Display::registryHandler(wl_registry *registry, uint32_t id,
 	}
-
-	if (interface == "wl_shell")
+	else if (interface == "wl_shell")
```
This and similar changes bellow are not described in the commit message and probably should not belong to this particular patch.

iusyk replied:
do you propose to have separate patch ?because , from my understanding, this approach is more clear and correct

lorc commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -498,7 +501,10 @@ void Display::init()
-	wl_registry_add_listener(mWlRegistry, &mWlRegistryListener, this);
+	if(-1 == wl_registry_add_listener(mWlRegistry, &mWlRegistryListener, this))
+	{
+		throw Exception("Can not add the listener.", errno);
```
This change is not described in the commit message.

Also, AFAIR, we don't use `if ( const == f() )` style.

iusyk replied:
replaced

lorc replied:
Commit should do one atomic change. In this case you clearly saying "that does not have a relation to the issue", so it is better to do in the separate commit. You can create that commit in this PR.

lorc reviewed (changes requested):
The same thing about commit description header.

For general approach check linux guidelines: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?h=v5.8-rc2

For instance, this one should look lile "Handle exeptions from callback functions".

Also, "cough" is a sudden expulsion of air through the large breathing passages. Probably you want "caught".

iusyk commented:
> The same thing about commit description header.
> 
> For general approach check linux guidelines: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?h=v5.8-rc2
> 
> For instance, this one should look lile "Handle exeptions from callback functions".
> 
> Also, "cough" is a sudden expulsion of air through the large breathing passages. Probably you want "caught".

fixed

lorc commented on src/displayBackend/wayland/Display.cpp:
```diff
@@ -408,8 +408,14 @@ void Display::sWaylandLog(const char* fmt, va_list arg)
+	}
+	catch(const std::exception& err)
+	{
+		LOG("Wayland", ERROR) << err.what();
```
Are you sure that this is a right way to handle the exception?
Looks like you just ignoring it.

I think, that we should properly handle it. Or stick to strategy "fail fast" and crash the program.

iusyk committed:
Replace the multi 'if; at 'if-else-if'

The code of Display::registryHandler has the multiple
blocks of if{}if{}.. All of the conditions are mutually exclusive.
The solution to optimize the code is to use the if-else-if blocks.

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>

rshym commented:
PR will be closed as outdated, as agreed with the author in the off-site discussion.
